manifests/extpack.pp

```diff
@@ -0,0 +1,87 @@
+# == Class: virtualbox::extpack
+#
+# This class (un)installs Oracle's VirtualBox extension pack.
+#
+# === Parameters
+#
+# [*ensure*]
+#   Set to 'present' to install extension pack. Set to 'absent' to uninstall.
+#   Defaults to 'present'
+#
+# [*version*]
+#   Full version of the Extension Pack to install. E.g. '4.3.20r96996'. This
+#   parameter is mandatory.
+#
+# [*source*]
+#   Download extension pack from the given URL. If not specified download from
+#   Oracle's official server.
+#
+class virtualbox::extpack (
+  $ensure  = $virtualbox::params::extpack_ensure,
+  $version = $virtualbox::params::extpack_version,
+  $source  = $virtualbox::params::extpack_source
+) {
+  validate_string($ensure)
+  validate_string($version)
+  validate_string($source)
+
+  if $version !~ /[^r]+r[^r]+/ {
+    fail("'${version}' is not a valid version string!")
+  }
+
+  case $ensure {
+    'present': {
+      if $virtualbox::package_ensure == 'absent' {
+        fail('Cannot install extension pack without installing VirtualBox')
+      }
+
+      # install/upgrade if we do not have the requested version
+      if $::virtualbox_extpack_version != $version {
+        $ver = regsubst($version, '^([^r]+)r([^r]+)', '\1')
+        $build = regsubst($version, '^([^r]+)r([^r]+)', '\2')
+        $filename = "Oracle_VM_VirtualBox_Extension_Pack-${ver}-${build}.vbox-extpack"
+        if $source != '' {
```

Comment:
Once we remove the `source` parameter (see my comment on the params class), we can change this logic to:

```ruby
if $source {
  ...
} else {
 ...
}
```